R/kenwardroger.R

```diff
@@ -0,0 +1,198 @@
+#'
+#' @param object (`mmrm`)\cr object.
+#' @param contrast (`numeric`)\cr contrast vector.
+#' @param linear (`logical`)\cr whether to use linear Kenward-Roger approximation.
+#'
+#' @return List with `est`, `se`, `df`, `t_stat` and `p_val`.
+#'
+#' @keywords internal
+h_df_1d_kr <- function(object, contrast, linear = FALSE) {
+  if (component(object, "reml") != 1) {
+    stop("Kenward-Roger is only for REML!")
+  }
+  assert_numeric(contrast, len = length(component(object, "beta_est")))
+  est <- sum(contrast * component(object, "beta_est"))
+  df <- h_kr_df(
+    v0 = object$beta_vcov,
+    l = matrix(contrast, nrow = 1),
+    w = component(object, "theta_vcov"),
+    p = object$kr_comp$P
+  )
+  se <- sqrt(h_quad_form_vec(contrast, object$beta_vcov_adj))
+
+  t_stat <- abs(est) / se
+  p_val <- pt(t_stat, df$m, lower.tail = FALSE) * 2
+
+  list(
+    est = est,
+    se = se,
+    df = df$m,
+    t_stat = t_stat,
+    p_val = p_val
+  )
+}
+
+#' Obtain the Adjusted Kenward-Roger Degree of Freedom
+#'
+#' @param v0 (`matrix`)\cr unadjusted covariance matrix.
+#' @param l (`matrix`)\cr linear combination matrix.
+#' @param w (`matrix`)\cr hessian matrix.
+#' @param p (`matrix`)\cr P matrix from `h_get_kr_comp`.
+#'
+#' @return Named list with elements:
+#' - `m`: `numeric` degree of freedom.
+#' - `lambda`: `numeric` f statistic scale parameter.
+#'
+#' @keywords internal
+h_kr_df <- function(v0, l, w, p) {
+  n_beta <- ncol(v0)
+  assert_matrix(v0, ncols = n_beta, nrows = n_beta)
+  assert_matrix(l, ncols = n_beta)
+  n_theta <- ncol(w)
+  assert_matrix(w, ncols = n_theta, nrows = n_theta)
+  n_visits <- ncol(p)
+  assert_matrix(p, nrows = n_visits * n_theta)
+  # see vignettes/kenward.Rmd#279
+  slvol <- solve(h_quad_form_mat(l, v0))
+  m <- h_quad_form_mat(t(l), slvol)
+  nl <- nrow(l)
+  mv0 <- m %*% v0
+  pl <- lapply(seq_len(nrow(p) / ncol(p)), function(x) {
+    ii <- (x - 1) * ncol(p) + 1
+    jj <- x * ncol(p)
+    p[ii:jj, ]
+  })
+  mv0pv0 <- lapply(pl, function(x) {
+    mv0 %*% x %*% v0
+  })
+  a1 <- 0
+  a2 <- 0
+  # see vignettes/kenward.Rmd#283
+  for (i in seq_len(length(pl))) {
+    for (j in seq_len(length(pl))) {
+      a1 <- a1 + w[i, j] * h_tr(mv0pv0[[i]]) * h_tr(mv0pv0[[j]])
+      a2 <- a2 + w[i, j] * h_tr(mv0pv0[[i]] %*% mv0pv0[[j]])
+    }
+  }
+  b <- 1 / (2 * nl) * (a1 + 6 * a2)
+  e <- 1 + a2 / nl
+  e_star <- 1 / (1 - a2 / nl)
+  g <- ((nl + 1) * a1 - (nl + 4) * a2) / ((nl + 2) * a2)
+  denom <- (3 * nl + 2 - 2 * g)
+  c1 <- g / denom
+  c2 <- (nl - g) / denom
+  c3 <- (nl + 2 - g) / denom
+  v_star <- 2 / nl * (1 + c1 * b) / (1 - c2 * b)^2 / (1 - c3 * b)
+  rho <- v_star / (2 * e_star^2)
+  m <- 4 + (nl + 2)  / (nl * rho - 1)
+  lambda <- m / (e_star * (m - 2))
+  list(m = m, lambda = lambda)
+}
+
+#' Obtain the Adjusted Covariance Matrix
+#'
+#' @param v (`matrix`)\cr unadjusted covariance matrix.
+#' @param w (`matrix`)\cr hessian matrix.
+#' @param p (`matrix`)\cr P matrix from `h_get_kr_comp`.
```

Comment:
```suggestion
#' @param p (`matrix`)\cr P matrix from [h_get_kr_comp()].
```
similar below. generally use markdown syntax for linking to other function docs